### PR TITLE
Fail build when incremental task does not declare outputs

### LINKED ISSUE #9723: Tasks without outputs shouldn't be up-to-date
The following task shouldn't be up-to-date after rerun, since it didn't declare any outputs:
```groovy
@Issue("https://github.com/gradle/gradle/issues/9723")
def "task without output is never up-to-date"() {
    def input = file('input.txt').createFile()
    buildFile << """
        class NoOutputTask extends DefaultTask {
            @InputFile File input
         
            @TaskAction execute(InputChanges inputChanges) {
            }
        }   
        
        task noOutput(type: NoOutputTask) {
            input = file('${input.name}')
        }
    """

    when:
    run 'noOutput'
    then:
    noneSkipped()

    when:
    run 'noOutput'
    then:
    noneSkipped()
}
```

as noted by @lptr, it should display the following:
```
> Task :noOutput
Caching disabled for task ':noOutput' because:
  Caching has not been enabled for the task
Task ':noOutput' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
:noOutput (Thread[Execution worker for ':',5,main]) completed. Took 0.002 secs.
```

### PR DESCRIPTION
**Pull request by wolfs**
Fixes #9723

See https://github.com/gradle/gradle-private/issues/2475

### CODE REVIEW

File: subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTasks.java [hunk 1]
```diff
@@ -130,6 +132,13 @@ public TaskWithIncrementalAction(Action<IncrementalTaskInputs> action) {
         public void doStuff(IncrementalTaskInputs changes) {
             action.execute(changes);
         }
+
+        @Optional
+        @OutputFiles
```
Comment (lptr) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTasks.java]: Why not use `@OutputFile` here?

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc [hunk 2]
```diff
@@ -127,6 +127,8 @@ Miscellaneous::
   Disable incremental Groovy compilation when you want to compile in those cases.
 * `ComponentSelectionRule` no longer can inject the metadata or ivy descriptor.
   Use the methods on the <<declaring_dependency_versions.adoc#sec:component_selection_rules,`ComponentSelection` parameter>> instead.
+* Declaring an <<custom_tasks.adoc#incremental_tasks,incremental task>> without declaring outputs is now deprecated.
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]: Is it deprecated or is it a failure?

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r51/TaskExecutionResultCrossVersionSpec.groovy [hunk 3]
```diff
@@ -124,16 +125,24 @@ class TaskExecutionResultCrossVersionSpec extends ToolingApiSpecification {
     def "reports incremental task as incremental"() {
         given:
         file('src').mkdir()
+        def supportsInputChanges = targetVersion > GradleVersion.version("5.4")
```
Comment (lptr) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r51/TaskExecutionResultCrossVersionSpec.groovy]: Nit: I think extracting the parameter type as a variable (and maybe the annotation, too) would be nicer.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTasks.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTasks.java]: No good reason. Changed to `@OutputFile` so it is in line with the name.

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc (hunk 2 above)
Comment (wolfs) [reply subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]: It is a failure. I changed the text.

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r51/TaskExecutionResultCrossVersionSpec.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r51/TaskExecutionResultCrossVersionSpec.groovy]: Done.

**Comment by lptr** [comment]
LGTM.

**Pull request merged**